plugins/methodinfo/src/main/java/org/qbicc/plugin/methodinfo/MethodDataEmitter.java

```diff
@@ -0,0 +1,387 @@
+            valueMap.put(sourceCodeInfoType.getMember(1), lf.literalOf(scInfo.getLineNumber()));
+            valueMap.put(sourceCodeInfoType.getMember(2), lf.literalOf(scInfo.getBcIndex()));
+            valueMap.put(sourceCodeInfoType.getMember(3), lf.literalOf(scInfo.getInlinedAtIndex()));
+            return lf.literalOf(sourceCodeInfoType, valueMap);
+        }).toArray(Literal[]::new);
+
+        return lf.literalOf(ts.getArrayType(sourceCodeInfoType, scInfoLiterals.length), List.of(scInfoLiterals));
+    }
+
+    Literal emitInstructionMap(CompilationContext ctxt, InstructionMap[] imapList) {
+        TypeSystem ts = ctxt.getTypeSystem();
+        LiteralFactory lf = ctxt.getLiteralFactory();
+        Section section = ctxt.getImplicitSection(ctxt.getDefaultTypeDefinition());
+
+        CompoundType.Member[] members = new CompoundType.Member[2];
+        int index = 0;
+        int compoundTypeSize = 0;
+        ValueType uint64Type = ts.getUnsignedInteger64Type();
+        members[0] = ts.getCompoundTypeMember("f" + (index++), uint64Type, compoundTypeSize, uint64Type.getAlign());
+        compoundTypeSize += uint64Type.getSize();
+        ValueType uint32Type = ts.getUnsignedInteger32Type();
+        members[1] = ts.getCompoundTypeMember("f" + (index++), uint32Type, compoundTypeSize, uint64Type.getAlign());
+        compoundTypeSize += uint32Type.getSize();
+
+        CompoundType imapType = ts.getCompoundType(CompoundType.Tag.STRUCT, "qbicc_instruction_map", compoundTypeSize, uint64Type.getAlign(), () -> List.of(members));
+
+        Literal[] imapLiterals = IntStream.range(0, imapList.length)
+            .parallel()
+            .mapToObj(i -> {
+            HashMap<CompoundType.Member, Literal> valueMap = new HashMap<>();
+            Function function = ctxt.getExactFunction(imapList[i].getFunction());
+            Literal functionCastLiteral = lf.bitcastLiteral(lf.literalOfSymbol(function.getName(), function.getType().getPointer()), ts.getUnsignedInteger8Type().getPointer());
+            Literal functionAddrLiteral = lf.valueConvertLiteral(lf.elementOfLiteral(functionCastLiteral, lf.literalOf(imapList[i].getOffset())), ts.getUnsignedInteger64Type());
+            valueMap.put(imapType.getMember(0), functionAddrLiteral);
+            valueMap.put(imapType.getMember(1), lf.literalOf(imapList[i].getSourceCodeIndex()));
+            section.declareFunction(null, function.getName(), function.getType());
+            return lf.literalOf(imapType, valueMap);
+        }).toArray(Literal[]::new);
+
+        return lf.literalOf(ts.getArrayType(imapType, imapLiterals.length), List.of(imapLiterals));
+    }
+
+    private void emitGlobalVariable(CompilationContext ctxt, String variableName, Literal value) {
+        Section section = ctxt.getImplicitSection(ctxt.getDefaultTypeDefinition());
+        section.addData(null, variableName, value);
+    }
+
+    public Path emitMethodData(CompilationContext ctxt, MethodData methodData) {
```

Comment:
The design of this metadata looks reasonable.  Can you describe it either in a comment here in the code or add a link to an issue that describes it?

Documenting the format and the reason for the layout choices pays off if it needs to be massaged in the future

Reply:
I would be adding structures for representing the same data at runtime (in future PRs). I think it would be better to add a reference to those structures here because it would be easier to understand the layout looking at those structures.

Structures would be something of this type:

```
    @internal
    public static final class method_info extends object {
        public uint32_t fileNameIndex;
        public uint32_t classNameIndex;
        public uint32_t methodNameIndex;
        public uint32_t methodDescIndex;
    }
```

Reply:
Can you describe the approach at a high-level in an issue?  Or in this PR?

If I'm reading the code correctly, a new `.ll` file is added called `__qbicc-default-type__` that contains this information for all functions.

This includes a single commonized string table with entries named as `.qbicc-md-string- + index`.
A `qbicc_method_info` table describing where methods come from.
A `qbicc_souce_code_info` which gives a mapping for bytecode index to linenumbers.
A `qbicc_instruction_map` that maps instructions to their corresponding source code info index.

I'm not 100% clear on which of these tables (and those emitted below with `emitGlobalVariable` are local to a function or global.

How much data is being generated here?  

Can you attach a copy of the `__qbicc-default-type__` file here?  It might be easier to look at the final output

Reply:
I have added some details in the [PR description](https://github.com/qbicc/qbicc/pull/594#issue-677342376). 
All these tables are treated as global data structures.

`__qbicc-default-type__.ll` is pretty big, around 53 MB. Corresponding object file is around 9 MB.
I am not able to add the file here. I can add the snippets of the tables. Hope that helps.

```
%T.struct._Unwind_Exception = type { i64, i8**, i64, i64 }
%T.struct._Unwind_Context = type { }
%T.struct.qbicc-md-string = type { i32, [4 x i8], i8* }
%T.struct.qbicc_instruction_map = type { i64, i64 }
%T.struct.qbicc_method_info = type { i32, i32, i32, i32 }
%T.struct.qbicc_souce_code_info = type { i32, i32, i32, i32 }

@.qbicc-md-string-data-0 = global [29 x i8] c"FormatterClosedException.java", align 1
@.qbicc-md-string-data-1 = global [19 x i8] c"FileDescriptor.java", align 1
@.qbicc-md-string-data-10 = global [9 x i8] c"Path.java", align 1
@.qbicc-md-string-data-100 = global [30 x i8] c"DefaultFileSystemProvider.java", align 1
@.qbicc-md-string-data-1000 = global [38 x i8] c"java/util/regex/PatternSyntaxException", align 1
@.qbicc-md-string-data-1001 = global [37 x i8] c"java/net/SocketPermissionCollection$1", align 1
@.qbicc-md-string-data-1002 = global [15 x i8] c"java/lang/Float", align 1
@.qbicc-md-string-data-1003 = global [52 x i8] c"java/util/concurrent/ConcurrentHashMap$ValueIterator", align 1
@.qbicc-md-string-data-1004 = global [15 x i8] c"java/lang/Class", align 1
...
@qbicc_classname_table = global [1332 x %T.struct.qbicc-md-string] [
 %T.struct.qbicc-md-string { i32 29, [4 x i8] zeroinitializer, i8* bitcast ([29 x i8]* @.qbicc-md-string-data-1867 to i8*) }, ; 0
 %T.struct.qbicc-md-string { i32 27, [4 x i8] zeroinitializer, i8* bitcast ([27 x i8]* @.qbicc-md-string-data-1872 to i8*) }, ; 1
 %T.struct.qbicc-md-string { i32 25, [4 x i8] zeroinitializer, i8* bitcast ([25 x i8]* @.qbicc-md-string-data-1877 to i8*) }, ; 2
 %T.struct.qbicc-md-string { i32 41, [4 x i8] zeroinitializer, i8* bitcast ([41 x i8]* @.qbicc-md-string-data-1882 to i8*) }, ; 3
 %T.struct.qbicc-md-string { i32 44, [4 x i8] zeroinitializer, i8* bitcast ([44 x i8]* @.qbicc-md-string-data-1885 to i8*) }, ; 4
 %T.struct.qbicc-md-string { i32 34, [4 x i8] zeroinitializer, i8* bitcast ([34 x i8]* @.qbicc-md-string-data-1891 to i8*) }, ; 5
...
@qbicc_filename_table = global [712 x %T.struct.qbicc-md-string] [
 %T.struct.qbicc-md-string { i32 12, [4 x i8] zeroinitializer, i8* bitcast ([12 x i8]* @.qbicc-md-string-data-380 to i8*) }, ; 0
 %T.struct.qbicc-md-string { i32 14, [4 x i8] zeroinitializer, i8* bitcast ([14 x i8]* @.qbicc-md-string-data-384 to i8*) }, ; 1
 %T.struct.qbicc-md-string { i32 22, [4 x i8] zeroinitializer, i8* bitcast ([22 x i8]* @.qbicc-md-string-data-389 to i8*) }, ; 2
 %T.struct.qbicc-md-string { i32 36, [4 x i8] zeroinitializer, i8* bitcast ([36 x i8]* @.qbicc-md-string-data-391 to i8*) }, ; 3
 %T.struct.qbicc-md-string { i32 19, [4 x i8] zeroinitializer, i8* bitcast ([19 x i8]* @.qbicc-md-string-data-395 to i8*) }, ; 4
 %T.struct.qbicc-md-string { i32 24, [4 x i8] zeroinitializer, i8* bitcast ([24 x i8]* @.qbicc-md-string-data-401 to i8*) }, ; 5
...
@qbicc_methodname_table = global [2748 x %T.struct.qbicc-md-string] [
 %T.struct.qbicc-md-string { i32 6, [4 x i8] zeroinitializer, i8* bitcast ([6 x i8]* @.qbicc-md-string-data-3595 to i8*) }, ; 0
 %T.struct.qbicc-md-string { i32 15, [4 x i8] zeroinitializer, i8* bitcast ([15 x i8]* @.qbicc-md-string-data-3599 to i8*) }, ; 1
 %T.struct.qbicc-md-string { i32 4, [4 x i8] zeroinitializer, i8* bitcast ([4 x i8]* @.qbicc-md-string-data-3602 to i8*) }, ; 2
 %T.struct.qbicc-md-string { i32 15, [4 x i8] zeroinitializer, i8* bitcast ([15 x i8]* @.qbicc-md-string-data-3607 to i8*) }, ; 3
 %T.struct.qbicc-md-string { i32 5, [4 x i8] zeroinitializer, i8* bitcast ([5 x i8]* @.qbicc-md-string-data-3613 to i8*) }, ; 4
 %T.struct.qbicc-md-string { i32 4, [4 x i8] zeroinitializer, i8* bitcast ([4 x i8]* @.qbicc-md-string-data-3619 to i8*) }, ; 5
...
@qbicc_method_info_table = global [8657 x %T.struct.qbicc_method_info] [
 %T.struct.qbicc_method_info zeroinitializer, ; 0
 %T.struct.qbicc_method_info { i32 0, i32 0, i32 2, i32 4 }, ; 1
 %T.struct.qbicc_method_info { i32 1, i32 1, i32 1, i32 7 }, ; 2
 %T.struct.qbicc_method_info { i32 1, i32 1, i32 3, i32 6 }, ; 3
 %T.struct.qbicc_method_info { i32 1, i32 1, i32 3, i32 5 }, ; 4
 %T.struct.qbicc_method_info { i32 1, i32 1, i32 1, i32 2 }, ; 5
...
@qbicc_source_code_info_table = global [68009 x %T.struct.qbicc_souce_code_info] [
 %T.struct.qbicc_souce_code_info { i32 5, i32 37, i32 3, i32 -1 }, ; 0
 %T.struct.qbicc_souce_code_info { i32 0, i32 1514, i32 7, i32 -1 }, ; 1
 %T.struct.qbicc_souce_code_info { i32 1, i32 1516, i32 4, i32 -1 }, ; 2
 %T.struct.qbicc_souce_code_info { i32 7, i32 37, i32 3, i32 -1 }, ; 3
 %T.struct.qbicc_souce_code_info { i32 6, i32 57, i32 1, i32 -1 }, ; 4
...
@qbicc_instruction_map_table = global [160067 x %T.struct.qbicc_instruction_map] [
 %T.struct.qbicc_instruction_map { i64 ptrtoint (i8* getelementptr ( i8, i8* bitcast (i8 addrspace(1)* (i8 addrspace(1)*, i8 addrspace(1)*)* @exact.java.util.HashMap$KeyIterator.next.ref.1.class.java-lang-Object.0 to i8*), i32 30) to i64), i64 66 }, ; 0
 %T.struct.qbicc_instruction_map { i64 ptrtoint (i8* getelementptr ( i8, i8* bitcast (void (i8 addrspace(1)*, i8 addrspace(1)*, i8 addrspace(1)*)* @init.java.util.HashMap$KeyIterator.1.ref.1.class.java-util-HashMap to i8*), i32 55) to i64), i64 1 }, ; 1
 %T.struct.qbicc_instruction_map { i64 ptrtoint (i8* getelementptr ( i8, i8* bitcast (i8 addrspace(1)* (i8 addrspace(1)*, i8 addrspace(1)*)* @exact.java.util.HashMap$KeyIterator.next.ref.1.class.java-lang-Object.0 to i8*), i32 76) to i64), i64 66 }, ; 2
 %T.struct.qbicc_instruction_map { i64 ptrtoint (i8* getelementptr ( i8, i8* bitcast (i8 addrspace(1)* (i8 addrspace(1)*, i8 addrspace(1)*)* @exact.java.util.HashMap$KeyIterator.next.ref.1.class.java-lang-Object.0 to i8*), i32 89) to i64), i64 66 }, ; 3
 %T.struct.qbicc_instruction_map { i64 ptrtoint (i8* getelementptr ( i8, i8* bitcast (void (i8 addrspace(1)*, i8 addrspace(1)*, i8 addrspace(1)*)* @init.java.util.HashMap$KeyIterator.1.ref.1.class.java-util-HashMap to i8*), i32 92) to i64), i64 258 }, ; 4
 %T.struct.qbicc_instruction_map { i64 ptrtoint (i8* getelementptr ( i8, i8* bitcast (void (i8 addrspace(1)*, i8 addrspace(1)*, i8 addrspace(1)*)* @init.java.util.HashMap$KeyIterator.1.ref.1.class.java-util-HashMap to i8*), i32 105) to i64), i64 258 }, ; 5
...